### PR TITLE
Connect Navigation page to live data

### LINKED ISSUE #4539: Manage the Navigation Menu Setting / Data
This is part of the Navigation epic: #4535, and requires #4537 

--- 

image

The list of navigation items should be stored as a JSON object in the settings table. 

This means creating a new [default setting](https://github.com/TryGhost/Ghost/blob/master/core/server/data/default-settings.json) called `navigation` with the type `blog`.
It should have one default item: "Home" - pointing at the root URL, which will be stored as just `/`.

On save, the values from the interface should be serialized into a JSON object and saved via the settings API.

The JSON object needs to contained ordered name and url pairs, something like:

```
{[
    {name: 'Home', url: '/'}
]}
```

* Pressing the `+` button should add another new row / item
* Pressing the save button should save all the rows which are not blank 
* Pressing the trash icon should remove the row / item
* Drag and drop reordering is dealt with in  #4540

I think the base URL should be stripped from the start of each URL when it is stored, so that we only store relative paths (and the navigation continues to function if the URL changes). If the URL doesn't start with the base URL, then we store and use the absolute URL.

### PR DESCRIPTION
**Pull request by jaswilli**
Closes #4539

### CODE REVIEW

File: core/client/controllers/settings/navigation.js [hunk 1]
```diff
@@ -1,23 +1,111 @@
...
+                    }
+                } else if (!validator.isURL(url) && url !== '' && url[0] !== '/') {
+                    url = '/' + url;
+                }
+
+                return {label: label, url: url};
+            }).compact();
+
+            this.set('model.navigation', JSON.stringify(navSetting));
+            this.get('model').notifyPropertyChange('navigation');
```
Comment (novaugust) [inline core/client/controllers/settings/navigation.js]: Ember doesn't notice this on its own? Huh, wonders why.

File: core/client/views/settings/navigation.js [hunk 2]
```diff
@@ -1,5 +1,12 @@
 import BaseView from 'ghost/views/settings/content-base';
 
-var SettingsNavigationView = BaseView.extend();
+var SettingsNavigationView = BaseView.extend({
+    keyDown: function (event) {
+        if (event.keyCode === 187) {
```
Comment (novaugust) [inline core/client/views/settings/navigation.js]: `=` adds an item?  Why not manage shortcuts with the router mixin?

File: core/client/controllers/settings/navigation.js (hunk 1 above)
Comment (jaswilli) [inline core/client/controllers/settings/navigation.js]: It does unless the new `model.navigation` property is the same as what was replaced.  The problem with that is that in the interim, the `navigationItems` computed property has been changed and may have the extra "blank" nav item in it, or the items may have additional properties (like `last`).

So to make sure that we start with a fresh `navigationItems` data structure after save, we have to give it a nudge.

File: core/client/views/settings/navigation.js (hunk 2 above)
Comment (jaswilli) [inline core/client/views/settings/navigation.js]: Yep.

File: core/client/controllers/settings/navigation.js (hunk 1 above)
Comment (novaugust) [inline core/client/controllers/settings/navigation.js]: :+1: I'm with you now. Seems like a good thing to leave a comment on, for people who bump into this down the road, to make sure they don't delete the line.

File: core/client/views/settings/navigation.js (hunk 2 above)
Comment (jaswilli) [inline core/client/views/settings/navigation.js]: Oh, just noticed the equal sign part.  That's an issue.  Needs to only fire on `+`.

Comment (novaugust) [inline core/client/views/settings/navigation.js]: Man, I thought `=` was a weird button to add stuff, but your
>Yep.

seemed so solid that I had no more doubts :D

**Comment by novaugust** [comment]
Looks good to me. It's currently possible to have Nav Items without a name or a url, does that need to be validated?

screen shot 2015-01-14 at 15 15 21

**Comment by jaswilli** [comment]
Yeah, I left it there on purpose as the only validation specified in the issue was to not save a row if it's entirely blank.  I suppose it could be valid to want to add all your Labels and then come back later or something...

I did have a ValidationEngine validation wired up for a while to do missing/too long validation on those fields but that opens another can of worms--mainly that this "Setting" model is one big record, so if you add validations, they're going to fire anywhere you try to save something stored in the setting record.  For example, go into Settings/General and make an invalid change, leave the screen and go to either Navigation and save or to Labs and toggle the Code Injection checkbox.

I was going to open a new issue about needing to handle this now that we're modifying settings in multiple places...

Anyway, tl;dr, validations haven't really been specced and for the above reason, I didn't want to create more weirdness with the way settings are saved.

**Comment by novaugust** [comment]
I'm happy to push it off to another issue =)
